src/librustc_lint/types.rs

```diff
@@ -344,11 +344,18 @@ impl LateLintPass for TypeLimits {
             }
         }
 
-        fn forbid_unsigned_negation(cx: &LateContext, span: Span) {
-            cx.sess()
-              .struct_span_err_with_code(span, "unary negation of unsigned integer", "E0519")
-              .span_help(span, "use a cast or the `!` operator")
-              .emit();
+        fn forbid_unsigned_negation(cx: &LateContext, span: Span, lit_span: Span) {
+            let mut err = cx.sess()
+                            .struct_span_err_with_code(span,
+                                                       "unary negation of unsigned integer",
+                                                       "E0519");
+              if let Ok(snip) = cx.sess().codemap().span_to_snippet(lit_span) {
+                err.span_suggestion(span, "try using a cast or the `!` operator:",
+                                    format!("!{}", snip));
```

Comment:
I think the cast way is the only one you can do generically. as `-1isize as usize` does not yield the same as `!1usize`. It would need to be `!(x-1)`, which errors if `x == 0`